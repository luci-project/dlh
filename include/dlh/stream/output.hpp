#pragma once

#include <dlh/syscall.hpp>
#include <dlh/stream/string.hpp>

template<size_t BUFFERSZ>
class OutputStream : public StringStream<BUFFERSZ> {
 protected:
	int fd;

 public:
	/*! \brief Default constructor  */
	explicit OutputStream(int fd) : fd(fd) {}

	/*! \brief Destructor */
	virtual ~OutputStream() {}

	/*! \brief Flush the buffer.
	 */
	virtual void flush() override {
		size_t l = 0;
		while (l < this->_pos) {
			auto r = Syscall::write(fd, this->_bufptr + l, this->_pos - l);
			if (!r.valid())
				break;
			else
				l += r.value();
		}
		this->_pos = 0;
	}

};

extern OutputStream<1024> cout, cerr;
