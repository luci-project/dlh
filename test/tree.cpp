#include <dlh/stream/output.hpp>
#include <dlh/container/tree.hpp>


int main(int argc, const char *argv[]) {
	cout << "TreeSet (using some integers):";
	TreeSet<int> s = { 888, 999 };
	s.insert(13);
	s.insert(3);
	s.insert(42);
	s.insert(23);
	s.insert(7);

	s.insert(13);
	s.insert(1549);
	s.emplace(666);
	s.insert(3085);
	s.insert(204);
	s.erase(666);

	s.erase(3085);
	s.insert(1572877);
	auto it1 = s.find(204);
	if (it1) {
		auto && tmp = s.extract(it1);
		tmp.value() = 205;
		s.insert(move(tmp));
	}
	s.insert(13);
	s.resize(128);
	s.insert(32);
	auto it2 = s.find(7);
	if (it2) {
		auto && tmp = s.extract(it2);
		tmp.value() = 13;
		s.insert(move(tmp));
	}
	s.insert(42);
	s.insert(52);

	#ifndef NDEBUG
	s.check();
	#endif

	TreeSet<int> t(s);
	for (auto i : t)
		cout << ' ' << i;
	cout << endl;

	int n = 42;
	cout << " - Lowest: " << *s.lowest() << endl;
	cout << " - Lower than " << n << ": " << *s.lower(n) << endl;
	cout << " - Lower than or equal to " << n << ": " << *s.floor(n) << endl;
	cout << " - Higher than or equal to " << n << ": " << *s.ceil(n) << endl;
	cout << " - Higher than " << n << ": " << *s.higher(n) << endl;
	cout << " - Highest: " << *s.highest() << endl;
	cout << endl << endl;


	cout << "TreeMap (using aruments)" << endl;
	TreeMap<const char *, int> m;
	for (int i = 0; i < argc; i++) {
		auto x = m.insert(argv[i], i);
		cout << " - insert " << argv[i] << " = " << x.second << endl;
		if (x.first)
			cout << "   * " << x.first->key << ", " << (*x.first).value << endl;
	}
	m["bar"] = 23;
	m["baz"] = -42;
	m["bar"] *= -1;

	auto i = m.find(StrPtr("foo"));
	if (i)
		cout << " - find: " << i->key << " = " << i->value << endl;
	auto j = m.erase(i);
	if (j)
		cout << " - erase: " << *j << endl;

	#ifndef NDEBUG
	m.check();
	#endif

	cout << "Full map:" << endl;
	for (auto i : m)
		cout << "	m[" << i.key << "] = " << i.value << endl;

	return 0;
}
