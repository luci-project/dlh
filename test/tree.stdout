TreeSet (using some integers): 3 13 23 32 42 52 205 888 999 1549 1572877
 - Lowest: 3
 - Lower than 42: 32
 - Lower than or equal to 42: 42
 - Higher than or equal to 42: 42
 - Higher than 42: 52
 - Highest: 1572877


TreeMap (using aruments)
 - insert ../test-tree = true
   * ../test-tree, 0
Full map:
	m[../test-tree] = 0
	m[bar] = -23
	m[baz] = -42
